### DISCUSSION TITLE
Some problems with CREATE statements parsing.

### DISCUSSION THREAD

**Discussion by LonwoLonwo** [JSQLParser/JSqlParser#1437] [Q&A]
Hello!

This week I got the task of obtaining metadata for the SQLite database. But this is a pretty specific database that does not want to return me metadata in the form of a table or something else. It remains only to parse tables DDL. And then I was faced with the fact that half of our standard Chinook.db tables do not want to be parsed correctly.

Maybe you can help me to figure out, what can I Do with these problems.

```
net.sf.jsqlparser.JSQLParserException: Encountered unexpected token: "CREATE" "CREATE"
    at line 1, column 1.

Was expecting one of:

    "("
    "ALTER"
    "CALL"
    "COMMENT"
    "COMMIT"
    "DECLARE"
    "DELETE"
    "DESCRIBE"
    "DROP"
    "EXEC"
    "EXECUTE"
    "EXPLAIN"
    "GRANT"
    "INSERT"
    "MERGE"
    "PURGE"
    "RENAME"
    "RESET"
    "ROLLBACK"
    "SAVEPOINT"
    "SET"
    "SHOW"
    "TRUNCATE"
    "UPDATE"
    "UPSERT"
    "USE"
    "VALUES"
    "WITH"
    <K_SELECT>

	at net.sf.jsqlparser.parser.CCJSqlParserUtil.parseStatement(CCJSqlParserUtil.java:190)
	at net.sf.jsqlparser.parser.CCJSqlParserUtil.parse(CCJSqlParserUtil.java:63)
	at net.sf.jsqlparser.parser.CCJSqlParserUtil.parse(CCJSqlParserUtil.java:38)
	at org.jkiss.dbeaver.ext.sqlite.model.SQLiteMetaModel.getTableDDL(SQLiteMetaModel.java:75)
```

![2021-12-09 18_14_13-Interactive Demo — JSQLFormatter 0 1 9 documentation — Mozilla Firefox](https://user-images.githubusercontent.com/45152336/145425230-a737e276-2d29-4f99-9101-789998efa99b.png)

![2021-12-09 18_14_55-Interactive Demo — JSQLFormatter 0 1 9 documentation — Mozilla Firefox](https://user-images.githubusercontent.com/45152336/145425234-f2e94fa5-3cf5-4dfd-8a64-232ec7c8cee2.png)

![2021-12-09 18_15_06-Interactive Demo — JSQLFormatter 0 1 9 documentation — Mozilla Firefox](https://user-images.githubusercontent.com/45152336/145425236-ee91aee3-1c17-4094-a912-ad37fe8c13d8.png)

![2021-12-09 18_15_21-Interactive Demo — JSQLFormatter 0 1 9 documentation — Mozilla Firefox](https://user-images.githubusercontent.com/45152336/145425239-8d5f813d-790f-4166-abe6-01c73ef8ca04.png)

I think, that the main issue can be with square brackets. But a special option in the demo version didn't help...

**Comment by LonwoLonwo** [comment]
@manticore-projects 
Hello! I hope you'll give me a hand :)

**Comment by manticore-projects** [comment]
> Hello!
> 
> This week I got the task of obtaining metadata for the SQLite database. 

Congrats! :-D

> But this is a pretty specific database that does not want to return me metadata in the form of a table or something else. It remains only to > parse tables DDL. And then I was faced with the fact that half of our standard Chinook.db tables do not want to be parsed correctly.

Please share one or 2 sample DDLs with me and I will see what I can do.
At a first glance, I am not sure if we support quoting withing the `CREATE TABLE` statement properly, but I would like to check this out.

Just share the ling from the JSQLFormatter UI please.

**Reply by manticore-projects** [reply]
Maybe also try something like this"
```sql
SELECT 
  m.name as table_name, 
  p.name as column_name
FROM 
  sqlite_master AS m
JOIN 
  pragma_table_info(m.name) AS p
ORDER BY 
  m.name, 
  p.cid
```

Of course I still will help out with the parser.

**Reply by manticore-projects** [reply]
Actually, this thing seems to be on me:

1) JSQLParser seems to work fine (with SquaredBracketQuotation applied)
2) JSQLFormatter also seems to work fine  (with SquaredBracketQuotation applied)

However, the SQL Editor used for the Online Demo does not respect `SquaredBracketQuotation` and so shows the error marker in the UI, even when the statement is parsed and formatted correctly.

Formatting still works, just ignore the Error Markers for the time being,
I would provide you with an update asap.

**Reply by LonwoLonwo** [reply]
DDL examples:

```sql
CREATE TABLE [Album]
(
    [AlbumId] INTEGER  NOT NULL,
    [Title] NVARCHAR(160)  NOT NULL,
    [ArtistId] INTEGER  NOT NULL, Column1 BLOB NULL,
    CONSTRAINT [PK_Album] PRIMARY KEY  ([AlbumId]),
    FOREIGN KEY ([ArtistId]) REFERENCES [Artist] ([ArtistId]) 
		ON DELETE NO ACTION ON UPDATE NO ACTION
);

CREATE TABLE [Employee]
(
    [EmployeeId] INTEGER  NOT NULL,
    [LastName] NVARCHAR(20)  NOT NULL,
    [FirstName] NVARCHAR(20)  NOT NULL,
    [Title] NVARCHAR(30),
    [ReportsTo] INTEGER,
    [BirthDate] DATETIME,
    [HireDate] DATETIME,
    [Address] NVARCHAR(70),
    [City] NVARCHAR(40),
    [State] NVARCHAR(40),
    [Country] NVARCHAR(40),
    [PostalCode] NVARCHAR(10),
    [Phone] NVARCHAR(24),
    [Fax] NVARCHAR(24),
    [Email] NVARCHAR(60),
    CONSTRAINT [PK_Employee] PRIMARY KEY  ([EmployeeId]),
    FOREIGN KEY ([ReportsTo]) REFERENCES [Employee] ([EmployeeId]) 
		ON DELETE NO ACTION ON UPDATE NO ACTION
);

CREATE TABLE dbeaver_corp_products_keys (
    id INTEGER NOT NULL PRIMARY KEY,
    uk NOT NULL UNIQUE,
    name TEXT,
    CHECK (name LIKE "%dbeaver%")
);


```

**Reply by LonwoLonwo** [reply]
> SELECT 
>   m.name as table_name, 
>   p.name as column_name
> FROM 
>   sqlite_master AS m
> JOIN 
>   pragma_table_info(m.name) AS p
> ORDER BY 
>   m.name, 
>   p.cid

Yes, we use sqlite_master in our SQLite queries, but, unfortunately, it returns columns, indexes, tables, triggers, but not the keys.

**Reply by manticore-projects** [reply]
Good evening.

The first 2 queries parse without any problem, when `SquaredBracketQuotation` is activated. (I will fix the hiccup in the SQL Editor Panel soonest, the Error Markers are incorrect).

However, that last query is simply not supported by JSQLParser regarding:

1) `uk NOT NULL UNIQUE` -- where is the column type?
2) `name TEXT,` -- quite sure that `name` is a reserved keyword and thus needs to be quoted
3) `CHECK (name LIKE "%dbeaver%")` -- `CHECK` is not allowed here

How many of those DDLs do you need to parse?

**Reply by manticore-projects** [reply]
This works:
```sql
CREATE TABLE dbeaver_corp_products_keys (
    id           INTEGER     NOT NULL PRIMARY KEY
    , uk         VARCHAR (1) NOT NULL UNIQUE
    , name       TEXT
    -- Unsupported:
    -- , CHECK (name LIKE "%dbeaver%")
)
;
```

**Reply by LonwoLonwo** [reply]
o! yes! "uk" was without data type, thanks.

And I need to parse any DDLs. To show users the right key objects.
Thanks for the help. I will try to use SquaredBracketQuotation in my situation.

**Comment by manticore-projects** [comment]
On your Database MetaData implementation, maybe have a look at [SQLSheet](https://github.com/panchmp/sqlsheet) which emulates an Excel Spreadsheet based database.
There I have built such JDBC compliant MetaData based on the Spread Sheets and Column information. You could take that as a template and add the Index/Foreign Key information accordingly.

https://github.com/panchmp/sqlsheet/blob/master/src/main/java/com/sqlsheet/XlsDatabaseMetaData.java

**Comment by manticore-projects** [comment]
Greetings.

I have fixed the Online SQL Formatter, you can [try your samples](http://jsqlformatter.manticore-projects.com/jsqlformatter/demo.html?args=-c%20MIJQogggKmAEUQEIBk4G0IBsBGBXAtgLqwAUAULJbBjgQJIAmxVLdAcjAOJggt9sB5KLDYBVZMgpUANNSgBLAC6YApsz4iAahBDAAEjtIBGAGwAGAJSVBwsRKmVZGAE6L5AZ0WN1ldlx4aVDYi4pIssgDGAPaYBAB2RoGwKAKISSx2YTKwwAJsAMpQIBB+1AAKANIA+lh4RA58ZSB0ALI6AJqwFWCdJNS19EywFg2yAGIC4HScbF09pP2uHl5DI0ngYzxgbMBg+YtunsR9LocrxGvpVHmwomUAItBwgrAQwFB0eQ3pN-dgqDARAJXu9PmwyCMANxkMigSCAhAodBgfAAB0wUQAnio1KQGmgUeisTjvOk-GBuLwksFMqNqMgAIaeNgM-C46naXQGXgkABMliCQhC9nC1DG8mczNZ7I0bE5+kMfIF1iFtNFaAUyhl6TlOgVPIAzJY6WgQCpUVFXO4oFEfKwOBSeCbEBLFAALe4MxTaliPGAfFpgE16CUqT3eu2+p4BoPqiAMBjOFTudyRjLy7mkADsxvVwCUmLTsozioALLnsmh8oovT7ArquWWK45qMAorg4opnIWrg39aRy5cW2gylFPAzMG2GHWtHrMyQjM3YE4ym6onEZ-wSzzeaWh8uxQyAB5F4tzxW7-dOFEM+SYU9BbekcxXnJ5QrFUoj6qEjHYtTfFQTStB0cy9NQv7EiopKvhMUwzGBCymualqKNatrDIBlAbFsOx7BBaJ-jixwEUS-4wVhGg3HcfrPMCbwfF8VzXLMfwAnRIKMeCUIwnCTzwEgqCwAw2AqAyABuKjOFU0TOKiVSos4UQMLgERoVUADWKiYu4eIsPIDDMeSlL8KqoSwMBbQgJ03TtHSuAaekT4LlYNLmaIbB0AAiqIsbZHE0qBDAAAaUB0voYDABUpABWysDIHQ3SwAARAApCJYmSc4qXJSMUJAA+).

Cheers
